app/components/user/TwoFactor/configure.js

```
@@ -0,0 +1,325 @@
+            </Button>
+          </Panel.Section>
+        </Panel>
+      );
+    }
+  }
+
+  componentDidMount() {
+    // Immediately go generate a token
+    commitMutation(this.props.relay.environment, {
+      mutation: graphql`
+        mutation configureCreateMutation($input: TOTPCreateInput!) {
+          totpCreate(input: $input) {
+            clientMutationId
+            provisioningUri
+            totp {
+              id
+            }
+          }
+        }
+      `,
+      variables: { input: {} },
+      onCompleted: this.handleCreateMutationComplete,
+      onError: this.handleCreateMutationError
+    });
+  }
+
+  componentWillUnmount() {
+    // When unmounting, if we haven't finished activating, try to delete the TOTP record
+    if (this.state.totpId && this.state.generatedTOTP && !this.state.activatedTOTP) {
+      commitMutation(this.props.relay.environment, {
+        mutation: graphql`
+          mutation configureDeleteMutation($input: TOTPDeleteInput!) {
+            totpDelete(input: $input) {
+              clientMutationId
+            }
+          }
+        `,
+        variables: { input: { id: this.state.totpId } }
+      });
+    }
+  }
+
+  handleCodeChange = (event) => {
+    this.setState({ totpToken: event.target.value });
+  };
+
+  handleCreateMutationComplete = (mutationResult: TOTPCreateReturnType) => {
+    this.setState({
+      generatingTOTP: false,
+      generatedTOTP: true,
+      totpId: mutationResult.totpCreate.totp.id,
+      provisioningUri: mutationResult.totpCreate.provisioningUri
+    });
+  };
+
+  handleCreateMutationError = (error) => {
+    if (error) {
+      if (error.source && error.source.type === GraphQLErrors.ESCALATION_ERROR) {
+        location.reload();
+      } else {
+        // if (error.source && error.source.type === GraphQLErrors.RECORD_VALIDATION_ERROR) {
+        //   this.setState({ errors: error.source.errors });
+        // } else {
+        alert(JSON.stringify(error));
```

We have the flash gear you can throw this error at (but I suspect you'd do that anyway)

---

I’m going to go review what errors are actually possible in this state, and enhance the form with being able to tell you if the code is wrong.